charms/reactive/altrelations.py

```diff
@@ -41,9 +49,574 @@ def from_name(cls, relation_name):
+        on a given endpoint, units may show up in this collection more than
+        once.
+        """
+        if self._all_units is None:
+            units = chain.from_iterable(rel.units for rel in self.relations)
+            self._all_units = CombinedUnitsView(units)
+        return self._all_units
+
+
+class Relation:
+    def __init__(self, relation_id):
+        self._relation_id = relation_id
+        self._relation_name = relation_id.split(':')[0]
+        self._application_name = None
+        self._units = None
+        self._data = None
+
+    @property
+    def relation_id(self):
+        """
+        This relation's relation ID.
+        """
+        return self._relation_id
+
+    @property
+    def relation_name(self):
+        """
+        This relation's relation name.
+
+        This will be the same as the
+        :class:`~charms.reactive.altrelations.Endpoint`'s relation name.
+        """
+        return self._relation_name
+
+    @property
+    def application_name(self):
+        """
+        The name of the remote application for this relation, or ``None``.
+
+        This is equivalent to::
+
+            relation.units[0].unit_name.split('/')[0]
+        """
+        if self._application_name is None and self.units:
+            self._application_name = self.units[0].unit_name.split('/')[0]
+        return self._application_name
+
+    @property
+    def units(self):
+        """
+        A list view of all the units on this relation.
+
+        This is actually a
+        :class:`~charms.reactive.altrelations.CombinedUnitsView`, so the units
+        will be in order by unit name, and you can access a merged view of all
+        of the units' data with ``self.units.received`` and
+        ``self.units.received_json``.  You should be very careful when using the
+        merged data collections, however, and consider carefully what will
+        happen when there are multiple remote units.  It is probabaly better to
+        iterate over each unit and handle its data individually.  See
+        :class:`~charms.reactive.altrelations.CombinedUnitsView` for an
+        explanation of how the merged data collections work.
+
+        The view can be iterated and indexed as a list, or you can look up units
+        by their unit name.  For example::
+
+            by_index = relation.units[0]
+            by_name = relation.units['unit/0']
+            assert by_index is by_name
+            assert all(unit is relation.units[unit.unit_name]
+                       for unit in relation.units)
+            print(', '.join(relation.units.keys()))
+        """
+        if self._units is None:
+            self._units = CombinedUnitsView([
+                RelatedUnit(self, unit_name) for unit_name in
+                sorted(hookenv.related_units(self.relation_id))
+            ])
+        return self._units
+
+    @property
+    def send_json(self):
```

Comment:
What do you think about refactoring `send_json -> send` and `send -> send_raw`? This will make it more clear that the json variant should be used unless you need lower-level access to be compatible with an existing interface.

(`to_publish` and `to_publish_raw` if you agree on that change too)

Reply:
+1 on both.